### PR TITLE
APM: basic tracer setup with Disruption and Reconcile spans

### PR DESCRIPTION
**Pull request by nathantournant**
## What does this PR do?

- [X] Adds new functionality
- [ ] Alters existing functionality
- [ ] Fixes a bug
- [ ] Improves documentation or testing

Please briefly describe your changes as well as the motivation behind them:
- Provides better observability for the chaos-controller, with a basic APM config setup and the optional Datadog provider. All formats are from Open Telemetry's libraries, allowing possibilities for other providers compatible with OTEL's [TracerProvider](https://pkg.go.dev/go.opentelemetry.io/otel/trace#TracerProvider) API.
- For the Datadog provider; already integrates with Logs (group all logs for a given Disruption Trace) and Profiler data (get Profiling for a given reconcile run)
- Current state: 
  - creates a span for the disruption and stores it in the controller watcher context (1 context/disruption)
  - for every reconcile, creates a child span, allowing to see every reconcile in the flame graph
  - does not support restart (see future updates), still limited in the amount of events recorded in the spans
- Future updates: 
  - trace context propagation through the CRD (store TraceID/SpanID as object annotations at creation and rebuild the "disruption" parent span context from there). This will notably enable tracing through controller restarts, and not use the (sort of) unfit context store.
  - pass trace context to the injector for logs and distributed tracing
  - add disruption config, more events and spans to the tracer
  - ...

Preview:
image

## Code Quality Checklist

- [X] The documentation is up to date.
- [X] My code is sufficiently commented and passes continuous integration checks.
- [X] I have signed my commit (see [Contributing Docs](../CONTRIBUTING.md)).

## Testing

- [X] I leveraged continuous integration testing
    - [X] by depending on existing `unit` tests or `end-to-end` tests.
    - [ ] by adding new `unit` tests or `end-to-end` tests.
- [X] I manually tested the following steps:
    - Create a disruption on the relevant cluster, delete it and check datadog for spans
    - [X] locally.
    - [X] as a canary deployment to a cluster.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -370,7 +373,37 @@ func main() {
...
 	}
 	// handle profiler sink close on exit
-	defer prfl.Stop()
+	defer func() {
+		logger.Infow("closing profiler sink client before exiting", "sink", prfl.GetSinkName())
+		prfl.Stop()
+	}()
+
+	// tracer sink
+	ts, err := tracer.GetSink(logger, tracertypes.SinkDriver(cfg.Controller.TracerSink))
```
Comment (luphaz) [inline main.go:382]: Prefer better naming other bad naming + comment
```suggestion
	tracerSink err := tracer.GetSink(logger, tracertypes.SinkDriver(cfg.Controller.TracerSink))
```

File: main.go [hunk 2]
```diff
@@ -370,7 +373,37 @@ func main() {
 		}
 	}
 	// handle profiler sink close on exit
-	defer prfl.Stop()
+	defer func() {
+		logger.Infow("closing profiler sink client before exiting", "sink", prfl.GetSinkName())
+		prfl.Stop()
```
Comment (luphaz) [inline main.go:378]:
```suggestion
		profilerSink.Stop()
```

File: main.go [hunk 3]
```diff
@@ -370,7 +373,37 @@ func main() {
...
+	}()
+
+	// tracer sink
+	ts, err := tracer.GetSink(logger, tracertypes.SinkDriver(cfg.Controller.TracerSink))
+	if err != nil {
+		logger.Errorw("error while creating profiler sink, switching to noop", "error", err)
+
+		ts, err = tracer.GetSink(logger, tracertypes.SinkDriverNoop)
+
+		if err != nil {
```
Comment (luphaz) [inline main.go:388]: I mean... Can this happen?

If not, maybe let's have a NewNoop that returns no error (all)

File: main.go [hunk 4]
```diff
@@ -370,7 +373,37 @@ func main() {
...
+		}
+	}
+	// handle tracer sink close on exit
+	defer func() {
+		logger.Infow("closing tracer sink client before exiting", "sink", ts.GetSinkName())
+
+		if err := ts.Stop(); err != nil {
+			logger.Errorw("error closing tracer sink client", "sink", ms.GetSinkName(), "error", err)
+		}
+	}()
```
Comment (luphaz) [inline main.go:399]: Not sure it makes sense to do this for a Noop sink?

File: main.go [hunk 5]
```diff
@@ -370,7 +373,37 @@ func main() {
...
+
+	// initiate Open Telemetry, set it up with the sink Provider
+	otel.SetTracerProvider(ts.GetProvider())
+
+	if ms.MetricRestart() != nil {
```
Comment (luphaz) [inline main.go:404]: Previously it was not done, what is this line for?

There is an error returned, retrieve it and log it please.
```suggestion
	if err := ms.MetricRestart()l; err != nil {
	  logger.Errorw("error sending MetricRestart", "sink", ms.GetSinkName(), "error" , err)
```

File: o11y/profiler/datadog/datadog.go [hunk 6]
```diff
@@ -15,13 +15,16 @@ type Sink struct{}
...
-	))
+	err := profiler.Start(
+		profiler.WithProfileTypes(
+			profiler.CPUProfile,
+			profiler.HeapProfile,
+			profiler.BlockProfile,
+			profiler.MutexProfile,
+			profiler.GoroutineProfile,
+		),
+		profiler.WithLogStartup(false),
```
Comment (luphaz) [inline o11y/profiler/datadog/datadog.go:26]: Why disabling this?

File: o11y/tracer/datadog/datadog.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+func (Sink) GetLoggableTraceContext(span trace.Span) []interface{} {
+	stringLogContext := []string{
+		"dd.trace_id", convertTraceID(span.SpanContext().TraceID().String()),
+		"dd.span_id", convertTraceID(span.SpanContext().SpanID().String()),
+	}
+
+	interfaceLogContext := make([]interface{}, len(stringLogContext))
+	for i, v := range stringLogContext {
+		interfaceLogContext[i] = v
+	}
```
Comment (luphaz) [inline o11y/tracer/datadog/datadog.go:44]: Just so you know the alternative exists, it's definitely great like you did:
```suggestion
	interfaceLogContext := make([]interface{}, 0, len(stringLogContext))
	for i, v := range stringLogContext {
		interfaceLogContext = append(interfaceLogContext, v)
	}
```

File: o11y/tracer/datadog/datadog.go [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+// GetSinkName returns the name of the sink (datadog)
+func (d Sink) GetSinkName() string {
+	return string(types.SinkDriverDatadog)
+}
+
+// convertTraceID converts a OTLP Trace/Span ID into a DataDog Trace/Span ID
+func convertTraceID(id string) string {
+	if len(id) < 16 {
```
Comment (luphaz) [inline o11y/tracer/datadog/datadog.go:66]: 🤔 And Datadog SDK does not provide anything to do such things for us?

File: controllers/cache_handler.go [hunk 9]
```diff
@@ -555,17 +569,33 @@ func (r *DisruptionReconciler) manageInstanceSelectorCache(instance *chaosv1beta
...
+	tpl, ok := r.CacheContextStore[disCacheHash]
+	if !ok {
+		return nil, fmt.Errorf("disruption has no cache created")
+	}
+
+	if tpl.Ctx == nil {
+		return nil, fmt.Errorf("disruption context is nil")
+	}
+
+	return tpl.Ctx, nil
```
[the remark this answers is not in the document]
Comment (nathantournant) [reply controllers/cache_handler.go:588]: It's a valid feeling; the more I think about this design the more I feel we could change it to something cleaner -- maybe using channels and better scope handling ? 
However the disruption context store is another topic we discussed when Dynamic Targeting was implemented and should be revisited after https://github.com/DataDog/chaos-controller/pull/681 is merged.
If you think the context usage for trace context as it is isn't good enough, I'll be happy to work on next iteration (trace context in the CRD) before releasing APM. https://github.com/DataDog/chaos-controller/pull/681 probably being merged soon might push it up anyway !

File: main.go (hunk 3 above)
Comment (nathantournant) [reply main.go:388]: > I mean... Can this happen?

As things are, no. This is more of a interface safety so we never end with a potentially untracked error if the `noop` sink ever returns an error -- were we to ever change that code. Happy to ignore that error if you think it's useless / too much / should be modified with the noop sink if needed.

> maybe let's have a NewNoop that returns no error (all)

I like the clarity of consistently using the same function (`tracer.GetSink(...)`) for every tracer init, wether it's the `noop` sink or another. I believe this is a small price for cleaner, more readable `tracer` package entry point.

File: controllers/cache_handler.go [hunk 10]
```diff
@@ -555,17 +569,33 @@ func (r *DisruptionReconciler) manageInstanceSelectorCache(instance *chaosv1beta
...
+
 // clearInstanceCache closes the context for the disruption-related cache and cleans the cancelFunc array (if it exists)
 func (r *DisruptionReconciler) clearInstanceSelectorCache(instance *chaosv1beta1.Disruption) {
-	disNamespacedName := types.NamespacedName{Name: instance.Name, Namespace: instance.Namespace}
-
-	disSpecHash, err := instance.Spec.HashNoCount()
+	disCacheHash, err := r.getContextStoreHash(instance)
 	if err != nil {
-		r.log.Errorf("error getting disruption hash")
+		r.log.Error(err)
```
[the remark this answers is not in the document]
Comment (nathantournant) [reply controllers/cache_handler.go:595]: so something like `r.log.With("err", err).Errorf("could not get disruption spec hash")` ?

File: o11y/tracer/datadog/datadog.go (hunk 8 above)
Comment (nathantournant) [reply o11y/tracer/datadog/datadog.go:66]: according to [this doc](https://docs.datadoghq.com/tracing/other_telemetry/connect_logs_and_traces/opentelemetry/?tab=go) it doesn't seem so 🤷🏻‍♂️ 
after investigation, I didn't find anything relevant in the ddtrace library :/

File: o11y/profiler/datadog/datadog.go (hunk 6 above)
Comment (nathantournant) [reply o11y/profiler/datadog/datadog.go:26]: the startup log does not go through our usual logger (and I couldn't parameter this) and is picked up as an error -- which is confusing. Same with the tracer !

**Review by luphaz — APPROVED** [review]

File: main.go (hunk 5 above)
Comment (nathantournant) [reply main.go:404]: It was done earlier in the file and was moved for readability. It sends a metric to notify the controller restart.

File: main.go (hunk 3 above)
Comment (nathantournant) [reply main.go:388]: For this PR I'll remove the noop sink error handling -- I get it is indeed a little bit much.
I'd love to discuss this with you and change the Factory pattern across metrics, profiler, tracer in a separate PR.

File: main.go (hunk 4 above)
Comment (nathantournant) [reply main.go:399]: It doesn't; however it's useful for the DD sink and potential others.

**Review by aymericDD — APPROVED** [review]
LGTM 👍 The traces are really useful for debugging and optimisations 👌

File: chart/values.yaml [hunk 11]
```diff
@@ -24,6 +24,7 @@ controller:
   enableObserver: true # enable observer on targets, notifying of target warning status and events
   metricsSink: noop
   profilerSink: noop
+  tracerSink: datadog
```
Comment (ptnapoleon) [inline chart/values.yaml:27]: Why not noop in this chart?

File: controllers/disruption_controller.go [hunk 12]
```diff
@@ -199,6 +215,11 @@ func (r *DisruptionReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 			r.handleMetricSinkError(r.MetricsSink.MetricDisruptionCompletedDuration(time.Since(instance.ObjectMeta.CreationTimestamp.Time), []string{"disruptionName:" + instance.Name, "namespace:" + instance.Namespace}))
 			r.emitKindCountMetrics(instance)
 
+			// close the ongoing disruption tracing Span
```
[the remark this answers is not in the document]
Comment (ptnapoleon) [reply controllers/disruption_controller.go:244]: > The Disruption span has to follow the lifecycle of the Disruption across multiple Reconciles, which makes things more difficult.

Is that true? I'm not terribly familiar with the best practices for spans, but a span that can last hours seems abnormal?

File: chart/values.yaml (hunk 11 above)
Comment (nathantournant) [reply chart/values.yaml:27]: fixed, my mistake

File: api/v1beta1/disruption_spancontext.go [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/propagation"
+)
+
+const (
+	annotationSpanContextKey = "SpanContext"
```
Comment (luphaz) [inline api/v1beta1/disruption_spancontext.go:17]: I would expect to have something scoped to chaos, like [here](https://github.com/DataDog/chaos-controller/blob/main/api/v1beta1/disruption_webhook.go#L52)

I'm wondering why aren't we not also scoping the user info to chaos 🤔 that's a bug IMO (another PR)

File: api/v1beta1/disruption_spancontext.go [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+)
+
+var (
+	// ErrNoSpanContext is the error returned when the annotation does not contains expected span context key
+	ErrNoSpanContext = fmt.Errorf("span context not found in disruption annotations")
+)
+
+// SpanContext extracts this disruption's span context, injects it in the given context, then returns it
+func (r *Disruption) ExtractSpanContext(ctx context.Context) (context.Context, error) {
```
Comment (luphaz) [inline api/v1beta1/disruption_spancontext.go:26]:
```suggestion
func (r *Disruption) SpanContext(ctx context.Context) (context.Context, error) {
```

**Pull request merged**